### PR TITLE
feat(optdepends): add array selection

### PR DESCRIPTION
**Pull request by Henryws**
## Purpose

Currently, `optdepends` can either be entirely selected or entirely deselected, leaving the user with no choice of what optional dependencies they want to install.

## Approach

Create a prompt that will ask the user which `optdepends` they want to install, or choosing ranges (`2-5`), while also keeping `Y` and `N` options as to select everything or nothing.

## Checklist

- [x] I confirm that I have read the [contributing guidelines](https://github.com/pacstall/pacstall/blob/develop/CONTRIBUTING.md), and this pull request is abiding by all the clauses stated in the guideline.

##### Using `-P`
image
##### Selecting deps
image
##### None
image
##### Input as non-numbers
image
##### Ranges
image

### CODE REVIEW

**Comment by Henryws** [comment]
I need to add out of bounds checkers.

**Review by wizard-28 — CHANGES_REQUESTED** [review]
image

If the total number of `optdepends` are more than a handful, then the `[1 2 3]` prompt will become very long. I suggest truncating the series like `[1 .. 20]`

**Comment by Henryws** [comment]
I couldn't find a package that has more than 6 packages, but I modified the neofetch package and duplicated a line
image

**Review by wizard-28 — CHANGES_REQUESTED** [review]
Document the function in the codebase.

**Review by wizard-28 — APPROVED** [review]

**Review by D-Brox — CHANGES_REQUESTED** [review]
Move `select_options()` to the main script so it can also be used on `upgrade`'s package selection and for repo selection on `search`

**Review by D-Brox — CHANGES_REQUESTED** [review]
Allow other formatting
image
image

**Comment by Henryws** [comment]
ok so first, `1..5` should be replaced with `1-5`, and second, you want comma separated delimiters as well?

**Comment by D-Brox** [comment]
> ok so first, `1..5` should be replaced with `1-5`, and second, you want comma separated delimiters as well?

@Henryws It's better to have something more robust to user errors

**Comment by Henryws** [comment]
I'll make `1..5` internally turn into `1-5` then.

File: misc/scripts/install-local.sh [hunk 1]
```diff
@@ -198,10 +200,15 @@ function makeVirtualDeb {
 			unset z
 			# tab over the next line
 			echo -ne "\t"
-			ask "Do you want to install the optional dependencies" Y
-			if [[ $answer -eq 1 ]]; then
-				for optdep in "${optdeps[@]}"; do
-					deps+=" ${optdep%%: *}"
+			select_options "Select optional dependencies to install" "${#optdeps[@]}"
+			choices=( $(cat /tmp/pacstall-select-options) )
```
Comment (hwittenborn) [inline misc/scripts/install-local.sh:204]: Is the output newline separated in the cat command? If so I'd use `mapfile` like that's bot's saying

File: misc/scripts/install-local.sh [hunk 2]
```diff
@@ -198,10 +200,15 @@ function makeVirtualDeb {
...
+			if [[ "${choices[0]}" != "n" ]]; then
+				for i in "${choices[@]}"; do
+					(( i-- ))
+					s="${optdeps[$i]}"
+					if [[ -n $s ]]; then
```
Comment (hwittenborn) [inline misc/scripts/install-local.sh:209]: Could this just use `"${#s}"` or something so the variable can be quoted? This probably ain't gonna cause issues regardless, but I think Bash best practices should still be followed.

File: pacstall [hunk 3]
```diff
@@ -238,6 +239,61 @@ function download() {
...
+			if [[ -z $input ]]; then
+				echo "Y"
+			fi
+			echo "$input"
+		fi
+	else
+		echo "Y"
+		input="Y"
+	fi
+	if [[ -z $input ]] || [[ $input =~ ^[Yy]$ ]]; then
```
Comment (hwittenborn) [inline pacstall:267]: Same quoting stuff as the above comment in this function.

File: pacstall [hunk 4]
```diff
@@ -238,6 +239,61 @@ function download() {
...
+						split_arr=(${i//-/ })
+						;;
+					*..*)
+						split_arr=(${i//../ })
+						;;
+					*)
+						select_options "$message" "$length"
+						;;
+				esac
+				out+=( $(seq -s ' ' "${split_arr[0]}" "${split_arr[1]}") )
```
Comment (hwittenborn) [inline pacstall:285]: This should use `mapfile` too, right?

**Review by hwittenborn — CHANGES_REQUESTED** [review]

File: misc/scripts/install-local.sh (hunk 1 above)
Comment (Henryws) [reply misc/scripts/install-local.sh:204]: It is space separated.

File: misc/scripts/install-local.sh (hunk 2 above)
Comment (Henryws) [reply misc/scripts/install-local.sh:209]: I believe variables inside `[[` don't need to be quoted.

File: pacstall (hunk 4 above)
Comment (Henryws) [reply pacstall:285]: This command will have predictable output so I don't think it's needed.

File: misc/scripts/install-local.sh (hunk 1 above)
Comment (hwittenborn) [reply misc/scripts/install-local.sh:204]: Fair enough

File: misc/scripts/install-local.sh (hunk 2 above)
Comment (hwittenborn) [reply misc/scripts/install-local.sh:209]: They probably don't *need* to, but I still think it'd be best if they were still

File: pacstall (hunk 4 above)
Comment (hwittenborn) [reply pacstall:285]: I still think the mapfile approach would be cleaner looking though:

```sh
mapfile -O "${#out[@]}" -t out < <(seq -s ' ' "${split_arr[0]}" "${split_arr[1]}")
```

Comment (Henryws) [reply pacstall:285]: > I still think the mapfile approach would be cleaner looking though:
> 
> ```sh
> mapfile -O "${#out[@]}" -t out < <(seq -s ' ' "${split_arr[0]}" "${split_arr[1]}")
> ```

That doesn't seem to look "cleaner" tbh

File: misc/scripts/install-local.sh (hunk 2 above)
Comment (Henryws) [reply misc/scripts/install-local.sh:209]: Our policy at pacstall is don't quote variables inside `[[ ]]` unless they somehow need to.

Comment (hwittenborn) [reply misc/scripts/install-local.sh:209]: Fair enough

**Review by D-Brox — APPROVED** [review]

File: pacstall (hunk 4 above)
Comment (hwittenborn) [reply pacstall:285]: It's just avoiding nested `()` stuff I guess, it won't bother me too much if it's not changed

**Pull request merged**